library/arith.lisp

```diff
@@ -480,6 +480,99 @@ The fields are defined as follows:
+    (cosh  (:a -> :a))
+    (tanh  (:a -> :a))
+    (asin  (:a -> :a))
+    (acos  (:a -> :a))
+    (atan  (:a -> :a))
+    (asinh (:a -> :a))
+    (acosh (:a -> :a))
+    (atanh (:a -> :a)))
+
+  (define-class (Exponentiable :a)
+    (**      (:a -> :a -> :a))
+    (logBase (:a -> :a -> :a)))
+
+  (declare sqrt ((Dividable Integer :a) (Exponentiable :a) => :a -> :a))
+  (define (sqrt x)
+    "Returns the square root of the input"
+    (** x (general/ (the integer 1) (the integer 2))))
+
+  (define-class ((Num :a) (Dividable :a :a) (Exponentiable :a) => Float :a)
+    (e  :a)
+    (pi :a))
+
+  (declare exp ((Float :f) (Exponentiable :f) => :f -> :f))
+  (define (exp x)
+    (** e x))
+
+  (declare log ((Float :f) (Exponentiable :f) => :f -> :f))
+  (define (log x)
+    (logBase e x))
+
+  (define-class ((Float :f) => RealFloat :f)
+    (rationalize (:f -> Fraction))))
+
+(cl:defun %define-unary-wrapper (return-type function cl:&optional (package :cl))
+  "defines a function which is a unary wrapper around a corresponding CL function"
+  (cl:let ((x      (cl:gensym))
+           (cl-fun (cl:find-symbol (cl:symbol-name function) package)))
+    (cl:unless cl-fun
+      (cl:error "Could not find symbol ~s in package ~s" function package))
+    `(define (,function ,x)
+       (lisp ,return-type (,x)
+         (,cl-fun ,x)))))
+
+(cl:defmacro %define-float-instances (coalton-type cl:&optional (package :cl))
+  "Defines the float instances for a lisp floating-point type"
+  (cl:let ((underlying-type (cl:find-symbol (cl:symbol-name coalton-type) package)))
+    (cl:unless underlying-type
+      (cl:error "Could not find symbol ~s in package ~s" coalton-type package))
+    `(coalton-toplevel
+       (define-instance (Trigonometric ,coalton-type)
+         ,(%define-unary-wrapper coalton-type 'sin   package)
+         ,(%define-unary-wrapper coalton-type 'cos   package)
+         ,(%define-unary-wrapper coalton-type 'tan   package)
+         ,(%define-unary-wrapper coalton-type 'sinh  package)
+         ,(%define-unary-wrapper coalton-type 'cosh  package)
+         ,(%define-unary-wrapper coalton-type 'tanh  package)
+         ,(%define-unary-wrapper coalton-type 'asin  package)
+         ,(%define-unary-wrapper coalton-type 'acos  package)
+         ,(%define-unary-wrapper coalton-type 'atan  package)
+         ,(%define-unary-wrapper coalton-type 'asinh package)
+         ,(%define-unary-wrapper coalton-type 'acosh package)
+         ,(%define-unary-wrapper coalton-type 'atanh package))
+
+       (define-instance (Exponentiable ,coalton-type)
+         (define (** base power)
+           (lisp ,coalton-type (base power)
+             (cl:expt base power)))
+         (define (logBase base number)
+           (lisp ,coalton-type (base number)
+             (cl:log number base))))
+
+       (define-instance (Float ,coalton-type)
+         (define e
+           (lisp ,coalton-type ()
+             (cl:coerce (cl:exp 1) ',underlying-type)))
```

Comment:
I would `coerce` the `1`, not the result; you'll lose precision

Reply:
done